### DISCUSSION TITLE
BreakMeshByBlockGenerator: how to split inner node among subdomains?

### DISCUSSION THREAD

**Discussion by favinom** [idaholab/moose#29585] [Q&A Meshing]
Hi all,

I am trying to use BreakMeshByBlockGenerator.
I have four blocks which are included in the domain. See picture.
If I call the BreakMeshByBlockGenerator the mid node is not split, which I understand from how it implemented.
Is there a way to also split that node?

<img width="598" alt="subdomain" src="https://github.com/user-attachments/assets/c38d4219-e51d-4979-ad1f-e97ec7f01bee" />

Thank you

**Comment by GiudGiud** [comment]
hello

yes you can join up two domains at the top or bottom, then split. (using a RenameBlockGenerator before the BreakMeshBYBlock)
Then split that domain again (using a ParsedSubdomainGenerator for example)

**Reply by favinom** [reply]
Many thanks for your answer.

Actually, this is what I tried
```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    nx = 8
    ny = 8
    dim = 2
  []
  [block1]
    type = SubdomainBoundingBoxGenerator
    block_id = 1
    bottom_left = '0.25 0.25 0.0'
    top_right =   '0.75  0.5  0.0'
    input = gen
  []
  [block2]
    type = SubdomainBoundingBoxGenerator
    block_id = 2
    bottom_left = '0.25 0.5 0.0'
    top_right =   '0.75 0.75  0.0'
    input = block1
  []
  [breakmesh]
    input = block2
    type = BreakMeshByBlockGenerator
    block_pairs = '1 2'
    split_interface = true
    #add_interface_on_two_sides = true
  []
  [block3]
    type = SubdomainBoundingBoxGenerator
    block_id = 3
    bottom_left = '0.25 0.25 0.0'
    top_right =   '0.5 0.75  0.0'
    input = breakmesh
  []
  [block4]
    type = SubdomainBoundingBoxGenerator
    block_id = 4
    bottom_left = '0.5 0.25 0.0'
    top_right =   '0.75 0.75  0.0'
    input = block3
  []
  [breakmesh2]
    input = block4
    type = BreakMeshByBlockGenerator
    block_pairs = '3 4'
    split_interface = true
    #add_interface_on_two_sides = true
  []
[]
```
but unfortunately after I rename the blocks, information of adjacent elements is lost and I cannot apply interface kernels.
Is there a way not to lose it?

**Reply by GiudGiud** [reply]
> I cannot apply interface kernels.

If you use the BreakMeshByBLock you cannot use interface kernels anymore. Interface kernels require a contiguous mesh. Breaking the mesh at boundaries is used for mortar / contact approaches

Are you just trying to get the interfaces added to the mesh? If so, you just need to use this:
https://mooseframework.inl.gov/source/meshgenerators/SideSetsBetweenSubdomainsGenerator.html

**Reply by favinom** [reply]
I understand. I am trying to apply interface kernels to a jump operator between a broken mesh.

I would need duplicated nodes in the mesh and discontinuous basis functions. Is there a way to do this in moose?

**Reply by GiudGiud** [reply]
Hello

We do support discontinuous galerkin. 

Do you really need the mesh to be broken? If it is going to be broken, InterfaceKernel is the wrong system to look at

**Reply by GiudGiud** [reply]
We can do DG with a contiguous mesh, and without duplicating the nodes. The nodal values can be different on each element

**Reply by favinom** [reply]
I would need the same structure of DG but just along some interfaces. I would need the interface integrals over \gamma in the formula at the beginning of page 19 of https://arxiv.org/pdf/2412.00443 .

Actually breaking the mesh and using interface kernels was a perfect combination.

How would you suggest to do it?

**Reply by GiudGiud** [reply]
In how many places do you need to do this?

If not too many I would recommend using two variables, a non broken mesh and an interface kernel. Having two variables makes it easy to have a discontinuity

**Reply by favinom** [reply]
Ideally, we would like to go for 100  interfaces.

The approach with two variables is doable if the interface cuts the whole domain, but I do not know if it is possible if the interface is included in the domain, as in the picture I posted above.

Maybe I can try with constraints?

However, I think the feature I need could be of interest also for someone else.

What is BreakBoundaryOnSubdomainGenerator doing?

Many thanks

**Reply by GiudGiud** [reply]
Are they ordered very regularly (like all aligned in 1D so you could alternate just two sets of variables) ?
Or is it somewhat random?

>  but I do not know if it is possible if the interface is included in the domain, as in the picture I posted above.

It is possible to block restrict in parts of the mesh, even if it's an inclusion and not spanning the entire domain.

> Maybe I can try with constraints?

Mortar constraints are definitely a solution here. If you split the domain (e.g. use the BreakXYZMeshGenerator, that creates a non-conformal mesh)

> What is BreakBoundaryOnSubdomainGenerator doing?

It does what the BreakMeshByBlock does but only along a given boundary, and creating different breakages (the breaks have non-broken nodes at each end) for each subdomain along that boundary
